cli/cdsctl/config.go

```diff
@@ -119,159 +119,252 @@ func withAllCommandModifiers() []cli.CommandModifier {
-					preargs = append(preargs, autoDiscoveredWorkflow)
-				}
-			}
-
-			*args = append(preargs, *args...)
-
-			return nil
+			return discoverConf(c.Ctx, args)
 		},
 	)
 }
 
-var (
-	autoDiscoveredProj     string
-	autoDiscoveredApp      string
-	autoDiscoveredWorkflow string
-)
-
-func discoverConf() error {
-	r, err := repo.New(".")
-	if err != nil {
-		return err
+func discoverConf(ctx []cli.Arg, args *[]string) error {
+	var needProject, needApplication, needWorkflow bool
+	var projectKey, applicationName, workflowName string
+
+	// populates needs an init values from args and ctx
+	mctx := make(map[string]cli.Arg, len(ctx))
+	for i, arg := range ctx {
+		mctx[arg.Name] = arg
+		switch arg.Name {
+		case _ProjectKey:
+			needProject = true
+			if i < len(*args) {
+				projectKey = (*args)[i]
+			}
+		case _ApplicationName:
+			needApplication = true
+			if i < len(*args) {
+				applicationName = (*args)[i]
+			}
+		case _WorkflowName:
+			needWorkflow = true
+			if i < len(*args) {
+				workflowName = (*args)[i]
+			}
+		}
 	}
 
-	if proj, _ := r.LocalConfigGet("cds", "project"); proj != "" {
-		//It's already configured
-		autoDiscoveredProj = proj
-		autoDiscoveredApp, _ = r.LocalConfigGet("cds", "application")
-		autoDiscoveredWorkflow, _ = r.LocalConfigGet("cds", "workflow")
+	// updates needs from values found in args
+	needProject = needProject && projectKey == ""
+	needApplication = needApplication && applicationName == ""
+	needWorkflow = needWorkflow && workflowName == ""
+
+	if !(needProject || needApplication || needWorkflow) {
 		return nil
 	}
 
-	fetchURL, err := r.FetchURL()
-	if err != nil {
-		return err
+	// try to find existing .git repository
+	var repoExists bool
+	r, err := repo.New(".")
+	if err == nil {
+		repoExists = true
 	}
 
-	name, err := r.Name()
-	if err != nil {
-		return err
+	// if repo exists ask for usage
+	if repoExists {
+		gitProjectKey, _ := r.LocalConfigGet("cds", "project")
+		gitApplicationName, _ := r.LocalConfigGet("cds", "application")
+		gitWorkflowName, _ := r.LocalConfigGet("cds", "workflow")
+
+		// if all needs were found in git do not ask for confirmation and use the config
+		needConfirmation := (needProject && gitProjectKey == "") || (needApplication && gitApplicationName == "") || (needWorkflow && gitWorkflowName == "")
+
+		if needConfirmation {
+			fetchURL, err := r.FetchURL()
+			if err != nil {
+				return err
+			}
+			name, err := r.Name()
+			if err != nil {
+				return err
+			}
+			repoExists = cli.AskForConfirmation(fmt.Sprintf("Detected repository as %s (%s). Is it correct?", name, fetchURL))
+		}
 	}
 
-	if cli.AskForConfirmation(fmt.Sprintf("Detected repository as %s (%s). Is it correct?", name, fetchURL)) {
-		projs, err := client.ProjectList(true, true, cdsclient.Filter{
-			Name:  "repo",
-			Value: name,
-		})
-		if err != nil {
-			return err
+	// if repo exists and is correct get existing config from it's config
+	if repoExists {
```

Comment:
why not in the same if block than above ?

Reply:
Because repoExists can be set to false if previous block, if usage not confirmed